### ISSUE TITLE
Open-AI real-time api (tts) support

### ISSUE THREAD

**Issue by stevenbaert** [open-webui/open-webui#5894]
Didn't find anything back (yet) on the forum/web, but it's an obvious question: is there a way to use the new open-ai real-time api in open-webui or are you planning to support it? 

Reference:
https://openai.com/index/introducing-the-realtime-api/

🙏

**Comment by thiswillbeyourgithub** [comment]
Hi @tjbck i saw [on hackernews](https://news.ycombinator.com/item?id=41743327) that the [livekits used to make](https://github.com/livekit/agents) the openai realtime api as well as [cerebras voice](https://cerebras.vercel.app/) seems to be open source.

They have tons of demos and code on their github. I think there must be a [llama-omni](https://github.com/ictnlp/LLaMA-Omni) implementation somewhere that would be a killer feature for open-webui!

**Comment by InventoCasa** [comment]
+1 for this!
Would be very nice if OpenWebUI would support the new gpt-4o-realtime-preview model!

**Comment by aguilarcarboni** [comment]
Bump! gpt-4o-realtime would be awesome, don't even get me started on Omni because I'd go crazy. Anybody got any time to implement this?

**Comment by jbaenaxd** [comment]
We totally need this, specially for folks in Europe, where ChatGPT didn't make available this service through their app/web for Plus/Team users without a VPN. Fortunately, it's available through the API, so we (europeans) could self-hosted ourselves to access this great tool.

**Comment by spammenotinoz** [comment]
> Didn't find anything back (yet) on the forum/web, but it's an obvious question: is there a way to use the new open-ai real-time api in open-webui or are you planning to support it?
> 
> Reference: https://openai.com/index/introducing-the-realtime-api/
> 
> 🙏

Careful what you wish for this model is quite expensive, you pay for audio input and output.
$100.00 / 1M input tokens
$200.00 / 1M output tokens

**Comment by Fusseldieb** [comment]
Would indeed be very cool! +1

**Comment by PhilosophiMoonbeam** [comment]
👀

**Comment by odellus** [comment]
Price has been lowered. Any interest in this now?

**Comment by ddwinhzy** [comment]
+1

**Comment by thiswillbeyourgithub** [comment]
I am absolutely convinced that the future is to use family multimodal models and they are more and more popular and smaller and smaller. The most recent best example I could find was the mini PCM series that exists in various sizes and is already sort of implemented in llama.cpp and ollama projects. [Here's the link](https://github.com/OpenBMB/MiniCPM-o). To me, it is a feature that is more and more needed and quite frankly the future of LLMs at least for the short and medium term. To be honest, I even do think that it should be a high priority along with stability and in contrast to, for example, making a desktop app.

**Comment by Fusseldieb** [comment]
> To be honest, I even do think that it should be a high priority

I would say so, too

> along with stability and in contrast to, for example, making a desktop app.

A desktop app is superfluous imo, as all webkit browsers allow to 'install' OI as a shortcut as if it were a native app, which is sufficient in 99% of cases. I even pinned mine on the taskbar and it opens in less than a second. Also, 'installed', it gets rid off the address bar and everything, just like a 'native app' would. 
Plus, at this point most 'native apps' are just a electron wrapper, which would only eat RAM for absolutely nothing. 

![Image](https://github.com/user-attachments/assets/ddb008b5-624d-4a7f-829f-1df472d6fc5c)

**Comment by thiswillbeyourgithub** [comment]
I agree. Personnaly for all my friends I created an account and openned the url via the browser and used the "add to homescreen" feature and nobody noticed it wasn't a regular app.

**Comment by thiswillbeyourgithub** [comment]
FYI there is currently a free demo at https://demo.ultravox.ai/ for ultravox models, it's really interesting and anyone who hasn't yet given a try to real time models should do so, it opens up a lot of interactivity usecase. And their 70B model is not that expensive either, $0.05/minute, so $3 per hour. Although I don't know if that counts if you are not speaking. Also by signing up on their website you get 30 minutes free without even having to enter a credit card. Although it's still lacking [quantization](https://github.com/fixie-ai/ultravox/pull/33) so somewhat out of reach from most customers for now.

I can't wait to be able to play with those on customer hardware and deploy it with open-webui!

**Comment by thiswillbeyourgithub** [comment]
If anyone wants to see a good implementation of realtime APIs, the devs at gradio made the great lib [fastrtc](https://github.com/gradio-app/fastrtc) which includes great examples.

For example [in about 150 lines of code you get the openai realtime api, UI included!](https://huggingface.co/spaces/fastrtc/talk-to-openai/blob/main/app.py)
Here is [the code for gemini voice chats](https://huggingface.co/spaces/fastrtc/talk-to-gemini/blob/main/app.py)

I'm not sure what would be the best way to include that into open-webui though. I think you'd have to hardcode openai and gemini into it because they are not available anywhere else (no openrouter for example) at least that I'm aware of. But there will be alternatives as time goes on but that's too bad to have to wait go knows how long.

There is a [cookbook](https://fastrtc.org/cookbook/) too.

**Comment by thiswillbeyourgithub** [comment]
FYI an enthusiastic OpenAI employee is making itself available to help projects if they have issues implementing support for the realtime API. Look for `sean` in [this hn thread](https://news.ycombinator.com/item?id=43762409)

**Comment by JamesClarke7283** [comment]
> > Didn't find anything back (yet) on the forum/web, but it's an obvious question: is there a way to use the new open-ai real-time api in open-webui or are you planning to support it?
> > Reference: https://openai.com/index/introducing-the-realtime-api/
> > 🙏
> 
> Careful what you wish for this model is quite expensive, you pay for audio input and output. $100.00 / 1M input tokens $200.00 / 1M output tokens

Indeed I agree, OpenAI are overpriced for their offerings nowdays(they dont have first-movers advantage anymore, you are paying more for the brand and ecosystem than the product itself, unless your usecase just to happens to match the product, you aren't getting the best value for money), but it's as they say, "a fool and their money are soon parted" (Supplimental evidence: ['"Inteligence" VS Price Comparison'](https://artificialanalysis.ai/models/gemini-2-5-pro?models=openai_o3&endpoints=openai_o4-mini%2Copenai_gpt-4-1-nano%2Copenai_gpt-4-1-mini%2Copenai_gpt-4-1%2Copenai_o3%2Copenai_o3-mini-high%2Copenai_gpt-4o-chatgpt-03-25%2Cmistral_magistral-small%2Cmistral_mistral-small-3-2%2Cmistral_magistral-medium%2Cdeepseek_deepseek-r1-05-28%2Canthropic_claude-4-sonnet-thinking%2Canthropic_claude-3-7-sonnet%2Canthropic_claude-4-opus%2Canthropic_claude-3-7-sonnet-thinking%2Canthropic_claude-4-sonnet%2Cgoogle_gemini-2-5-flash-05-20-reasoning_ai-studio%2Cgoogle_gemini-2-5-pro-06-05_ai-studio%2Cgoogle_gemini-2-5-flash-lite-reasoning_ai-studio%2Cfireworks_qwen3-235b-a22b-instruct-reasoning%2Ccerebras_qwen3-235b-a22b-instruct-reasoning%2Ccerebras_qwen3-32b-instruct-reasoning%2Cbaseten_kimi-k2%2Cxai_grok-4#intelligence-vs-price) ), it's diminishing returns for most people(including myself), i gave up OpenAI ages ago.

People are sceptical and dont have the tokens to burn for local testing.

Side Rant: Training Compute isn't the deciding asset for entering frontier status anymore folks (I know i am preach to the choir here), my opinion will be, next it will be 'corporate private/internal data' that loses the competivive advantage(OSINT + grounded synthetic/augmented data), there is only so much a single company can do longterm to improve a ML based chatbot before a new paradigm is needed(the elusive 'World Models'?).
For best quality ML insight, HuggingFace is currently the platform to be i find (as of July 2025), even as a novice, there are open source 'spaces'(Mini-Apps) for almost every usecase.

Blind human evaluation driven ELO leaderboard here(they may be near top, but observe how the gap's get smaller & smaller):
https://lmarena.ai/leaderboard



The data transparency is a little spotty as only samples of the data get released infrequently, but it tends to line up with my human preferences(if you factor in the 90% CI in your list of models to 'vibe check' for own usecase). 
It's also i think, a catch-22 for gaming benchmarks as traditional benchmarks and live, blind human preference evaluations cannot easily be gamed simultaniously(apart from freak incidents like social engineering attacks to admin's like Meta-LLAMA did with llama4 by doing the old-bait-and-switch, and got called out for it).

Externally sourced Evaluations/Benchmarks are only meant to help people choose a set models to try out of course(so they don't overwhelm themselves or... their wallet).

From my personal usage and mistakes(by spending too much with OpenAI), I find using a Frontier Open Weight model (Currently R1 & Qwen) from OpenRouter(Remote Inference), or better(locally) is much better, you can also finetune LoRa's easier and easier now as a novice (Soon LocalAI will have a OpenAI complient training API, after which it's just a question of using a open source frontend for doing OpenAI API integrated training). 

The lower sized Qwen3 series (32B/30-a3b) work well for me & uncensored/ablated models are typically better for nuanced topics anyway if you use them responsibly, especially for people where biased answers are a productivity burden (i.e think academic/corporate research/knowledge-base curation, anything where post-training bias can skew insights too much).

Bottom line:
If/When OpenWebUI implements this feature, it's support for OpenAI API 'RealTime' spec, so you would be able to use it with any OpenAI API inference engine server that supports 'Realtime'. including Open Source engines.


like for example:
https://github.com/theboringhumane/echoOLlama
or more definitively for a self-contained suite: LocalAI has an issue here:
https://github.com/mudler/LocalAI/issues/3714

End result of having this in the inference engines like LocalAI (which is making progress) and then supporting it in the UI like with OpenWebUI voice mode, would be you could save money and regain freedom/control in the process.

The Realtime API spec is one of the reasons OpenAI's Advanced voice is so smooth and natural (relatively speaking).

The inference server  can use a chained approach (STT -> LLM -> TTS), which is at a technical level a bit harder, as information can be lost unless you encode the STT incantations,subtitles to Text, and reverse for TTS.

End goal is using open weight speech-to-speech (Omni) models, but for now i think its pretty awesome you can swap out STT/LLM/TTS to make one with OpenWebUI.

A universal REST api standard like "OpenAI API spec" just makes convergence much simpler(the inference server is abstracted away, one can use OpenAI or Local, or someone else's), much like with MCP, but without the [glaring security issues](https://www.trolleyesecurity.com/articles-news-mcp-servers-exposed-without-any-security/) with its core design (I.E: core defect transcends software implementation).

**Comment by Mcayear** [comment]
Any progress?

**Comment by Classic298** [comment]
+1

**Comment by Classic298** [comment]
the gemini-live-2.5-flash-preview-native-audio models are INSANELY good and understand even words that are spoken but not in the dictionary.

Unfortunately, these are NOT simple TTS or STT models, these are multi modal models that want direct audio input.

Same with gpt-4o-realtime models

they want direct input.

**Comment by yashsinha-aero** [comment]
Is there any update to this? Any way to use Gemini’s Live API?

**Comment by ghost** [comment]
it's nice to support real-time API and automatically falls back to TTS+STT when real-time is not supported in the remote endpoint.

[livekit](https://livekit.io/) integration would also be nice

**Comment by Classic298** [comment]
@swiftdev29 no update feel free to PR

**Comment by rbb-dev** [comment]
Hey
putting this here in case any dev wants to take a look
I’ve got a working fork of Open WebUI with OpenAI realtime voice support, and I’d really like people to try it and report whatever is wrong with it

docker:
ghcr.io/rbb-dev/open-webui-realtime:latest

package:
https://github.com/rbb-dev/open-webui/pkgs/container/open-webui-realtime

docs:
https://github.com/rbb-dev/docs/tree/realtime-voice-docs

It supports live realtime voice, typed handoff into the same session, tool execution during calls, and continuity/history merged back into normal chat.

If anything breaks, feels weird, or looks badly thought through, please let me know.

**Comment by spammenotinoz** [comment]
Excellent, if the voices were not fixed, I could test with more providers
Will monitor the cost, but excellent work.  It's so much better than the standard voice.

**Comment by rbb-dev** [comment]
As long as the base URL is correct and other providers support a full realtime API along with ephemeral key minting, it should function properly.
However, it will only work with the realtime models, as the system specifically looks for "realtime" in the model name.
Currently, OpenAI API does not distinguish realtime models in the /models endpoint any differently, so this is the only available method for now.
Thanks for the feedback @spammenotinoz, let me know if you encounter any issues.
if you encounter something, i'll need OWUI logs with "GLOBAL_LOG_LEVEL=DEBUG" set.

**Comment by rbb-dev** [comment]
@spammenotinoz few bugs were fixed, update your docker image when you have time,

https://github.com/rbb-dev/open-webui/commit/fc2456b46295d2c05af3ddb6cfd5369222f83b8d
https://github.com/rbb-dev/open-webui/commit/041312905aba34405af3b65fb532259eaca2b03b

**Comment by LauraGPT** [comment]
For a self-hosted real-time STT backend, [FunASR](https://github.com/modelscope/FunASR) (16K+ stars) now provides an **OpenAI-compatible `/v1/audio/transcriptions` endpoint**:

```bash
pip install funasr fastapi uvicorn python-multipart
python examples/openai_api/server.py --device cuda --port 8000
```

Since Open WebUI already supports OpenAI STT, you can point it to `http://localhost:8000/v1` as a local alternative.

**Advantages:**
- 170x realtime on GPU (vs Whisper 13x)
- 50+ languages, auto-detection
- Speaker diarization built-in
- WebSocket streaming for real-time voice input
- No API key, no costs, MIT license

Also has a Docker streaming server for real-time WebSocket:
```bash
docker run -p 10095:10095 registry.cn-hangzhou.aliyuncs.com/funasr_repo/funasr:funasr-runtime-sdk-online-cpu-0.1.12
```

**Comment by tq1008** [comment]
OpenAI real-time API support would be huge. The WebRTC-based approach with `input_audio_buffer.append` + `response.create` is quite different from the standard chat completions streaming model.

In the meantime, for anyone who wants lower-latency streaming *today*: you can already get sub-20ms TTFT with any OpenAI-compatible streaming endpoint that has servers in Asia. The standard `/v1/chat/completions` with `stream: true` + SSE gives you incremental token delivery that feels near-real-time for text, even if it is not the full WebRTC audio pipeline.

If you are in Southeast Asia, picking a provider with Singapore or Tokyo PoPs makes a noticeable difference vs. US-west endpoints (200ms+ RTT).

**Comment by Pl8tinium** [comment]
> Hey putting this here in case any dev wants to take a look I’ve got a working fork of Open WebUI with OpenAI realtime voice support, and I’d really like people to try it and report whatever is wrong with it
> 
> docker: ghcr.io/rbb-dev/open-webui-realtime:latest
> 
> package: https://github.com/rbb-dev/open-webui/pkgs/container/open-webui-realtime
> 
> docs: https://github.com/rbb-dev/docs/tree/realtime-voice-docs
> 
> It supports live realtime voice, typed handoff into the same session, tool execution during calls, and continuity/history merged back into normal chat.
> 
> If anything breaks, feels weird, or looks badly thought through, please let me know.

what i found is that it cant invoke tools from the model, the native tool integration of openwebui doesnt seem to be integrated yet, correct?

**Comment by rbb-dev** [comment]
No, it has complete tools integration on par with any other model.

<img width="1075" height="467" alt="Image" src="https://github.com/user-attachments/assets/ad275773-e357-4463-8892-27a959fcd2f2" />

**Comment by rbb-dev** [comment]
Did you enabled native tools in the models advanced settings?

**Comment by Pl8tinium** [comment]
are you refering to this section of the model parameters? yes here i enabled the default web search and the OR web search. Unfortunately doesnt work with either of them. 

<img width="1026" height="1034" alt="Image" src="https://github.com/user-attachments/assets/18796c15-14cf-4159-920a-e3384c09ed91" />

though it seems to be able to call other kinds of tools as it can be seen in this screenshot

<img width="1395" height="1308" alt="Image" src="https://github.com/user-attachments/assets/4ee5d59d-fc40-4edc-9876-e355d097b8cb" />

**Comment by SarcasticBaka29** [comment]
@rbb-dev Hey there! I'm trying out HF's  [fully local speech-to-speech pipeline ](https://github.com/huggingface/speech-to-speech) and it seems to work pretty darn well. Do you think it would be possible to integrate into your open-webui fork since it exposes an OpenAI compatible /v1/realtime endpoint?
